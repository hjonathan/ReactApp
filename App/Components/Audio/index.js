import React, {Component} from "react";
import {Platform, View, Text, TouchableOpacity, ProgressBarAndroid} from "react-native";
import moment from "moment";
import Sound from "react-native-sound";
import AudioRecorder from "react-native-audio-record";
import MaterialIcon from "react-native-vector-icons/MaterialIcons";
import I18n from "../../I18n";
import * as Res from "../../Assets/resources";
import Styles from "./styles";

const PREFIX = "formsliderAud";
let timer = null;
/**
 * Class Audio
 * Example:
 *      ...
 *      import Audio from "../../Components/Audio";
 *
 *      class SomeComponent extends component {
 *          render () {
 *              return (
 *                  <Audio
 *                      onClose={()=>{}} // close function
 *                      saveAudio={() => {}} // saving function
 *                  />
 *              );
 *          }
 *      }
 */
class Audio extends Component {
    /**
     * Constructor
     * @param props
     */
    constructor (props) {
        super(props);
        this.state = {
            timer: null,
            minutes: "00",
            seconds: "00",
            time: 0.00,
            audioPath: null,
            stopped: false,
            loading: false,
            recording: false
        };
    }

    /**
     * Initializes audio options
     */
    componentDidMount () {
        const options = {
            sampleRate: 16000,
            channels: 1,
            bitsPerSample: 16,
            wavFile: this.nameGenerator()
        };
        AudioRecorder.init(options);
    }
    /**
     * Closes control panel
     */
    onClosePanel () {
        this.setState({
            time: 0,
            audioPath: null,
            loading: false
        });
        this.props.onClose({status: false});
    }

    /**
     * Generates a file name
     * @returns {string}
     */
    nameGenerator () {
        let date = moment().format("YYYYMMDD"),
            timestamp = moment().valueOf(),
            extension = "m4a",
            uid;
        uid = `${PREFIX}_${date}_${timestamp}.${extension}`;
        return uid;
    }

    /**
     * Records audio
     * @private
     */
    _record () {
        if (!this.state.recording && !this.state.loading) {
            this.timeStart();
            AudioRecorder.start();
            this.setState({
                stopped: false,
                recording: true,
                loading: true
            });
        }
    }

    /**
     * Stops the current recording
     * @private
     */
    async _stop () {
        if (this.state.recording) {
            this.timeClear();
            let audioFile = await AudioRecorder.stop();
            this.setState({
                stopped: true,
                recording: false,
                loading: false,
                audioPath: audioFile
            });
        }
    }

    /**
     * Plays the current record
     * @private
     */
    _play () {
        if (this.state.stopped && !this.state.recording) {
            const sound = new Sound(this.state.audioPath, null, (error) => {
                if (error) {
                    console.error(error);
                    return;
                }
                this.setState({loading: true});
                this.timeStart();
                sound.play((success) => {
                    this.setState({loading: !success});
                    this.timeClear();
                });
            });
        }
    }

    /**
     * Saves current record
     * @private
     */
    _save () {
        if (this.state.stopped && this.state.audioPath) {
            this.props.saveAudio({path: this.state.audioPath});
        }
    }

    /**
     * Start time of chronometer
     */
    timeStart = () => {
        timer = setInterval(() => {
            let {minutes} = this.state,
                seconds = (Number(this.state.seconds) + 1).toString();
            if (Number(this.state.seconds) === 59) {
                minutes = (Number(this.state.minutes) + 1).toString();
                seconds = "00";
            }
            this.setState({
                minutes: minutes.length === 1 ? `0${minutes}` : minutes,
                seconds: seconds.length === 1 ? `0${seconds}` : seconds
            });
        }, 1000);
        this.setState({timer});
    };

    /**
     * End time of chronometer
     */
    timeClear () {
        this.setState({
            timer: null,
            minutes: "00",
            seconds: "00"
        });
        clearInterval(timer);
    }

    /**
     * Renders progress bar
     * @returns {*}
     */
    progressBar () {
        if (Platform.OS === "android") {
            return (
                <ProgressBarAndroid styleAttr="Horizontal" color={Res.colors.teal} />
            );
        }
        return null;
    }

    /**
     * Renders header
     * @returns {*}
     */
    renderHeader () {
        return (
            <View style={Styles.headerContainer}>
                <View style={Styles.headerTitle}>
                    <Text style={Styles.textTitle}>
                        {I18n.t("selectAudioRecordDialogtitle")}
                    </Text>
                </View>
                <TouchableOpacity
                    onPress={() => {
                        this.onClosePanel();
                    }}
                >
                    <MaterialIcon name="clear" style={Styles.closeIcon} />
                </TouchableOpacity>
            </View>
        );
    }

    /**
     * Renders indeterminate progressBar and timer
     * @returns {*}
     */
    renderProgressBar () {
        if (this.state.loading) {
            return (
                <View style={{backgroundColor: Res.colors.white}}>
                    {this.progressBar()}
                    <View>
                        <Text>{`${this.state.minutes}:${this.state.seconds}`}</Text>
                    </View>
                </View>
            );
        }
        return null;
    }

    /**
     * Renders record button
     * @returns {*}
     */
    renderRecordButton () {
        return (
            <View style={Styles.buttonContainer}>
                <TouchableOpacity
                    onPress={() => {
                        this._record();
                    }}
                >
                    <View style={Styles.button}>
                        <MaterialIcon name="fiber-manual-record" style={Styles.buttonIcon} />
                    </View>
                </TouchableOpacity>
                <Text>{I18n.t("recordAudioStringDialog")}</Text>
            </View>
        );
    }

    /**
     * Renders stop button
     * @returns {*}
     */
    renderStopButton () {
        return (
            <View style={Styles.buttonContainer}>
                <TouchableOpacity
                    onPress={() => {
                        this._stop();
                    }}
                >
                    <View style={Styles.button}>
                        <MaterialIcon name="stop" style={Styles.buttonIcon} />
                    </View>
                </TouchableOpacity>
                <Text>{I18n.t("stopStringAudioDialog")}</Text>
            </View>
        );
    }

    /**
     * Renders play button
     * @returns {*}
     */
    renderPlayButton () {
        return (
            <View style={Styles.buttonContainer}>
                <TouchableOpacity
                    onPress={() => {
                        this._play();
                    }}
                >
                    <View style={Styles.button}>
                        <MaterialIcon name="play-arrow" style={Styles.buttonIcon} />
                    </View>
                </TouchableOpacity>
                <Text>{I18n.t("playStringAudioDialog")}</Text>
            </View>
        );
    }

    /**
     * Renders save button
     * @returns {*}
     */
    renderSaveButton () {
        return (
            <View style={[Styles.buttonContainer, {margin: 4}]}>
                <TouchableOpacity
                    onPress={() => {
                        this._save();
                    }}
                >
                    <View style={Styles.button}>
                        <MaterialIcon name="send" style={Styles.buttonIcon} />
                    </View>
                </TouchableOpacity>
                <Text>{I18n.t("savedStringAudioDialog")}</Text>
            </View>
        );
    }

    /**
     * Renders component
     * @returns {*}
     */
    render () {
        return (
            <View style={Styles.mainContainer}>
                <View style={Styles.panelContainer}>
                    {this.renderHeader()}
                    {this.renderProgressBar()}
                    <View style={Styles.subContainer}>
                        {this.renderRecordButton()}
                        {this.renderStopButton()}
                        {this.renderPlayButton()}
                        {this.renderSaveButton()}
                    </View>
                </View>
            </View>
        );
    }
}

export default Audio;
