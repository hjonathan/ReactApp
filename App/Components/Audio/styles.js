import {StyleSheet} from "react-native";
import * as Res from "../../Assets/resources";

export default StyleSheet.create({
    mainContainer: {
        flex: 1,
        alignItems: "center",
        justifyContent: "center"
    },
    panelContainer: {
        width: "70%",
        flexDirection: "column",
        justifyContent: "center"
    },
    headerContainer: {
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "space-between",
        backgroundColor: Res.colors.lightColor,
        height: 38,
        padding: 8
    },
    headerTitle: {
        flex: 1,
        justifyContent: "center"
    },
    textTitle: {
        textAlign: "center",
        color: Res.colors.white
    },
    closeIcon: {
        color: Res.colors.white,
        fontSize: 28
    },
    subContainer: {
        flexDirection: "row",
        flexWrap: "wrap",
        justifyContent: "center",
        backgroundColor: Res.colors.white,
        padding: 16
    },
    buttonContainer: {
        flexWrap: "wrap",
        alignItems: "center",
        width: "33%"
    },
    button: {
        borderRadius: 8,
        padding: 4,
        backgroundColor: Res.colors.mainColor
    },
    buttonIcon: {
        color: Res.colors.white,
        fontSize: 36
    }
});
