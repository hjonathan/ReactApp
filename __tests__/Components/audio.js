import 'react-native';
import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import Audio from '../../App/Components/Audio';

jest.mock('react-native-sound', () => require.requireActual('../../__mocks__/react-native-sound').default);
jest.mock('react-native-audio-record', () => require.requireActual('../../__mocks__/react-native-audio-record').default);

describe('Component >>> audio', () => {
    const onClose = () => {},
        saveAudio = () => {};

    let audioView,
        audioViewDefault;

    beforeEach(() => {
        audioViewDefault = shallow(<Audio />);
        audioView = shallow(<Audio onClose={onClose} saveAudio={saveAudio}/>);
    });

    test('Render audio', () => {
        expect(toJson(audioView)).toMatchSnapshot();
        expect(toJson(audioViewDefault)).toMatchSnapshot();
    });

    test('Render renderHeader', () => {
        let header,
            close;
        header = audioView.instance().renderHeader();
        close = header.props.children[1].props;
        expect(close.onPress()).toMatchSnapshot();
    });

    test('Render progressBarAndroid', () => {
        jest.mock('Platform', () => {
            return {OS: 'android'};
        });
        expect(audioView.instance().progressBar()).toMatchSnapshot();
        jest.mock('Platform', () => {
            return {OS: 'ios'};
        });
        expect(audioView.instance().progressBar()).toMatchSnapshot();
    });

    test('Render progressBar&Timer', () => {
        let progressBarTimer;
        audioView.instance().setState({loading: true});
        progressBarTimer = audioView.instance().renderProgressBar();
        expect(progressBarTimer).toMatchSnapshot();
    });

    test('Render recordButton', () => {
        let recordButton;
        recordButton = audioView.instance().renderRecordButton();
        audioView.instance().setState({recording: false, loading: false});
        expect(recordButton.props.children[0].props.onPress()).toMatchSnapshot();
        audioView.instance().setState({recording: true, loading: true});
        expect(recordButton.props.children[0].props.onPress()).toMatchSnapshot();
    });

    test('Render stopButton', async () => {
        let stopButton;
        stopButton = audioView.instance().renderStopButton();
        audioView.instance().setState({recording: true});
        expect(await stopButton.props.children[0].props.onPress()).toMatchSnapshot();
        audioView.instance().setState({recording: false});
        expect(await stopButton.props.children[0].props.onPress()).toMatchSnapshot();
    });

    test('Render playButton', () => {
        let playButton;
        playButton = audioView.instance().renderPlayButton();
        audioView.instance().setState({stopped: true, recording: false});
        expect(playButton.props.children[0].props.onPress()).toMatchSnapshot();
        audioView.instance().setState({stopped: false, recording: false});
        expect(playButton.props.children[0].props.onPress()).toMatchSnapshot();
    });

    test('Render saveButton', () => {
        let saveButton;
        saveButton = audioView.instance().renderSaveButton();
        audioView.instance().setState({stopped: true, audioPath: "path"});
        expect(saveButton.props.children[0].props.onPress()).toMatchSnapshot();
        audioView.instance().setState({stopped: false, audioPath: ""});
        expect(saveButton.props.children[0].props.onPress()).toMatchSnapshot();
    });

    test('timeStart function', () => {
        audioView.instance().timeStart();
    });

    test('timeClear function', () => {
        audioView.instance().timeClear();
    });
});
