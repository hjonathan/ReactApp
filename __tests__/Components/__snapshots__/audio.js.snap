// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component >>> audio Render audio 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Component
    style={
      Object {
        "flexDirection": "column",
        "justifyContent": "center",
        "width": "70%",
      }
    }
  >
    <Component
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#3397E1",
          "flexDirection": "row",
          "height": 38,
          "justifyContent": "space-between",
          "padding": 8,
        }
      }
    >
      <Component
        style={
          Object {
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#FFF",
              "textAlign": "center",
            }
          }
        >
          selectAudioRecordDialogtitle
        </Text>
      </Component>
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
      >
        <Icon
          allowFontScaling={false}
          name="clear"
          size={12}
          style={
            Object {
              "color": "#FFF",
              "fontSize": 28,
            }
          }
        />
      </TouchableOpacity>
    </Component>
    <Component
      style={
        Object {
          "backgroundColor": "#FFF",
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "center",
          "padding": 16,
        }
      }
    >
      <Component
        style={
          Object {
            "alignItems": "center",
            "flexWrap": "wrap",
            "width": "33%",
          }
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
        >
          <Component
            style={
              Object {
                "backgroundColor": "#1F6CBC",
                "borderRadius": 8,
                "padding": 4,
              }
            }
          >
            <Icon
              allowFontScaling={false}
              name="fiber-manual-record"
              size={12}
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 36,
                }
              }
            />
          </Component>
        </TouchableOpacity>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          recordAudioStringDialog
        </Text>
      </Component>
      <Component
        style={
          Object {
            "alignItems": "center",
            "flexWrap": "wrap",
            "width": "33%",
          }
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
        >
          <Component
            style={
              Object {
                "backgroundColor": "#1F6CBC",
                "borderRadius": 8,
                "padding": 4,
              }
            }
          >
            <Icon
              allowFontScaling={false}
              name="stop"
              size={12}
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 36,
                }
              }
            />
          </Component>
        </TouchableOpacity>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          stopStringAudioDialog
        </Text>
      </Component>
      <Component
        style={
          Object {
            "alignItems": "center",
            "flexWrap": "wrap",
            "width": "33%",
          }
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
        >
          <Component
            style={
              Object {
                "backgroundColor": "#1F6CBC",
                "borderRadius": 8,
                "padding": 4,
              }
            }
          >
            <Icon
              allowFontScaling={false}
              name="play-arrow"
              size={12}
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 36,
                }
              }
            />
          </Component>
        </TouchableOpacity>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          playStringAudioDialog
        </Text>
      </Component>
      <Component
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexWrap": "wrap",
              "width": "33%",
            },
            Object {
              "margin": 4,
            },
          ]
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
        >
          <Component
            style={
              Object {
                "backgroundColor": "#1F6CBC",
                "borderRadius": 8,
                "padding": 4,
              }
            }
          >
            <Icon
              allowFontScaling={false}
              name="send"
              size={12}
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 36,
                }
              }
            />
          </Component>
        </TouchableOpacity>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          savedStringAudioDialog
        </Text>
      </Component>
    </Component>
  </Component>
</Component>
`;

exports[`Component >>> audio Render audio 2`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Component
    style={
      Object {
        "flexDirection": "column",
        "justifyContent": "center",
        "width": "70%",
      }
    }
  >
    <Component
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#3397E1",
          "flexDirection": "row",
          "height": 38,
          "justifyContent": "space-between",
          "padding": 8,
        }
      }
    >
      <Component
        style={
          Object {
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#FFF",
              "textAlign": "center",
            }
          }
        >
          selectAudioRecordDialogtitle
        </Text>
      </Component>
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
      >
        <Icon
          allowFontScaling={false}
          name="clear"
          size={12}
          style={
            Object {
              "color": "#FFF",
              "fontSize": 28,
            }
          }
        />
      </TouchableOpacity>
    </Component>
    <Component
      style={
        Object {
          "backgroundColor": "#FFF",
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "center",
          "padding": 16,
        }
      }
    >
      <Component
        style={
          Object {
            "alignItems": "center",
            "flexWrap": "wrap",
            "width": "33%",
          }
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
        >
          <Component
            style={
              Object {
                "backgroundColor": "#1F6CBC",
                "borderRadius": 8,
                "padding": 4,
              }
            }
          >
            <Icon
              allowFontScaling={false}
              name="fiber-manual-record"
              size={12}
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 36,
                }
              }
            />
          </Component>
        </TouchableOpacity>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          recordAudioStringDialog
        </Text>
      </Component>
      <Component
        style={
          Object {
            "alignItems": "center",
            "flexWrap": "wrap",
            "width": "33%",
          }
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
        >
          <Component
            style={
              Object {
                "backgroundColor": "#1F6CBC",
                "borderRadius": 8,
                "padding": 4,
              }
            }
          >
            <Icon
              allowFontScaling={false}
              name="stop"
              size={12}
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 36,
                }
              }
            />
          </Component>
        </TouchableOpacity>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          stopStringAudioDialog
        </Text>
      </Component>
      <Component
        style={
          Object {
            "alignItems": "center",
            "flexWrap": "wrap",
            "width": "33%",
          }
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
        >
          <Component
            style={
              Object {
                "backgroundColor": "#1F6CBC",
                "borderRadius": 8,
                "padding": 4,
              }
            }
          >
            <Icon
              allowFontScaling={false}
              name="play-arrow"
              size={12}
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 36,
                }
              }
            />
          </Component>
        </TouchableOpacity>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          playStringAudioDialog
        </Text>
      </Component>
      <Component
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexWrap": "wrap",
              "width": "33%",
            },
            Object {
              "margin": 4,
            },
          ]
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
        >
          <Component
            style={
              Object {
                "backgroundColor": "#1F6CBC",
                "borderRadius": 8,
                "padding": 4,
              }
            }
          >
            <Icon
              allowFontScaling={false}
              name="send"
              size={12}
              style={
                Object {
                  "color": "#FFF",
                  "fontSize": 36,
                }
              }
            />
          </Component>
        </TouchableOpacity>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          savedStringAudioDialog
        </Text>
      </Component>
    </Component>
  </Component>
</Component>
`;

exports[`Component >>> audio Render playButton 1`] = `undefined`;

exports[`Component >>> audio Render playButton 2`] = `undefined`;

exports[`Component >>> audio Render progressBar&Timer 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#FFF",
    }
  }
>
  <Component>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      00:00
    </Text>
  </Component>
</Component>
`;

exports[`Component >>> audio Render progressBarAndroid 1`] = `null`;

exports[`Component >>> audio Render progressBarAndroid 2`] = `null`;

exports[`Component >>> audio Render recordButton 1`] = `undefined`;

exports[`Component >>> audio Render recordButton 2`] = `undefined`;

exports[`Component >>> audio Render renderHeader 1`] = `undefined`;

exports[`Component >>> audio Render saveButton 1`] = `undefined`;

exports[`Component >>> audio Render saveButton 2`] = `undefined`;

exports[`Component >>> audio Render stopButton 1`] = `undefined`;

exports[`Component >>> audio Render stopButton 2`] = `undefined`;
